x-pack/plugins/observability/server/routes/copilot/route.ts

```diff
@@ -40,12 +42,91 @@ const promptRoutes: {
           throw Boom.notImplemented();
         }
 
-        return client.chatCompletion.create(prompt.messages(resources.params.body as any));
+        try {
+          return await client.chatCompletion.create(prompt.messages(resources.params.body as any));
+        } catch (error: any) {
+          if (axios.isAxiosError(error) && error.response?.status === 401) {
+            throw Boom.forbidden(error.response?.statusText);
+          }
+          throw error;
+        }
       },
     });
   })
 );
 
+const trackRoute = createObservabilityServerRoute({
+  endpoint: 'POST /internal/observability/copilot/prompts/{promptId}/track',
+  params: t.type({
+    path: t.type({
+      promptId: t.string,
+    }),
+    body: t.intersection([
+      t.type({
+        responseTime: t.number,
+        messages: t.array(
+          t.intersection([
+            t.type({
+              role: t.union([
+                t.literal(ChatCompletionRequestMessageRoleEnum.System),
+                t.literal(ChatCompletionRequestMessageRoleEnum.User),
+                t.literal(ChatCompletionRequestMessageRoleEnum.Assistant),
+              ]),
+              content: t.string,
+            }),
+            t.partial({
+              name: t.string,
+            }),
+          ])
+        ),
+        response: t.string,
+      }),
+      t.partial({
+        feedbackAction: t.union([t.literal('thumbsup'), t.literal('thumbsdown')]),
+      }),
+    ]),
+  }),
+  options: {
+    tags: [],
```

Comment:
I think that's OK for now, given the tech preview status and us trying to figure out ownership.

Reply:
Would protecting this endpoint with AccessTags cause any foreseeable issues? If not, I think securing it would be the best way forward.

Reply:
do you have any suggestion?

Reply:
One last question: Are the calls to the 3rd party endpoints in this feature free or paid? Since there isn't authorization, would there be a potential concern of anyone calling these endpoints and racking up the bill?

Reply:
Users have to configure credentials for their own (Azure) OpenAI deployment, so that's their responsibility - for the S3 endpoint, we're footing the bill, but we don't expect a huge amount of requests here.

Reply:
I would suggest creating a new sub-feature that includes new access tags, these tags could be called `copilot_read`/`copilot_write` or you could include those tags into an existing Observability feature. Then add these tags to your routes

Reply:
I would consider that out of scope for this, as it's a tech preview and opt-in, and we're likely to remove some of this stuff in 8.10. Also we're past feature freeze, this is really just about logging the feedback.

Reply:
IMO security of our API endpoints should be in scope. The tech preview designation doesn't absolve us of our security responsibilities. Would you be opposed to adding API access tags to the existing Observability features defined in `x-pack/plugins/observability/server/plugin.ts`, and updating the routes here with those access tags? That would allow any existing Observability user to call these endpoints, and it'll give you the ability to add/remove from this feature set in a BWC manner without locking yourself into a specific privilege model.

Reply:
I see there's about a dozen endpoints in several Observability plugins that have an empty array for `options.tags`. Not saying that's an excuse to not do it, but I'm curious when it is required and when it is not? If the answer to the latter is "never", does it make sense to add a runtime check for this? If I add access tags, can I remove them in 8.10 without causing a breaking change? To clarify, this feature can only be enabled via kibana.yml and comes with big disclaimers (for various reasons).

Reply:
is the idea to add (for instance) a `copilot` access tag to all these features?

<img width="724" alt="CleanShot 2023-07-05 at 20 23 46@2x" src="https://github.com/elastic/kibana/assets/352732/aac3d4dd-848d-4c1e-a3a8-086700d7b2fd">

Reply:
> Not saying that's an excuse to not do it, but I'm curious when it is required and when it is not? If the answer to the latter is "never", does it make sense to add a runtime check for this? If I add access tags, can I remove them in 8.10 without causing a breaking change? To clarify, this feature can only be enabled via kibana.yml and comes with big disclaimers (for various reasons).

This is something we are actively working on defining alongside the product security team, but my short answer is:
1) Routes that CRUD saved objects in order to function typically don't need access tags, as we will already be performing authorization checks when you interact with the SO service. Derivatives of the SO service are also sufficient, such as the services required to interact with Cases and Alerts.
2) Routes that perform meaningful actions against the cluster (writing data, performing an RPC style call, or returning data to the client `asInternalUser`) should have an access tag applied.
3) Routes that interact with third-party services should have an access tag applied
4) Routes that incur a non-insignificant load on the Kibana server should have an access tag applied.

By default, all you need to call a Kibana API is a valid set of credentials. A username/password with no roles assigned is sufficient today. That's typically not what we want to happen, but we also don't want to incur unnecessary round-trips to ES to authorize each and every call that we make. That's why we rely on the SO service for many routes, because that'll be performing authorization anyway, so adding an additional authorization step before that is redundant.

tl;dr (with caveats): If your route uses the SO service early in its route handler to do _something_, then you probably don't need to add an access tag. Otherwise, you should consider having an access tag.

We don't want every engineer to have to know these detais, and we're discussing ways that we can simplify our authorization model so that we are secure by default, without performing more `_has_privileges` checks than necessary.

Reply:
> is the idea to add (for instance) a copilot access tag to all these features?

Yes, that was my suggestion. It's based on my weak assumption that the AI Assistant in question is available to all of these features, but I defer to you on that

Reply:
Thanks Larry! I'll link to your description in the team chat as well, it's useful for all of us to know I think. I'll add the access tags then, will be fun with how many CODEOWNERS reviews that will trigger 😬